mantidproject/mantid
# Engin-X: more flexible bank interpretation

Fixes #11808.

This adds aliases `North` and `South` for the input property `Bank`, and a new input property `DetectorIndices` that can be used instead of `Bank` to have full control over what detectors are used for calibration.

**To test:**
- Tests should go well
- Code review and maybe some manual checks (see below)

With this you can do Bank='1' as before, but also DetectorIndices='0-1199' or DetectorIndices='0-99, 1000-1100, 10000-10100'. This new option could be called DetectorIndices or WorkspaceIndices (to avoid confusion with the detector IDs), as they are actually the workspace indices as it stands now. This is something we'll need to check with instrument scientists, as the detector IDs that we have in the ENGIN-X IDF are a bit non-intuitive and the last day Saurabh was not very sure about this. 

Because the new 'North' and 'South' aliases for banks 1 and 2, the property Bank is now of string type. And also, Bank doesn't have a default value now, as you can use either Bank or DetectorIndices.

For validation I've included a few unit test cases where the results when using Bank='1' and Bank='2' are compared with the results when using the equivalent North/South alias and ranges of indices.
You can also run manually some commands like in the tests, like:

```
ws=Load("ENGINX00228061.nxs", OutputWorkspace='ENGIN-X_test_ws')
# all 4 are equivalent:
EnginXFocus(InputWorkspace=ws, Bank='1', OutputWorkspace='out_1')
EnginXFocus(InputWorkspace=ws, Bank='North', OutputWorkspace='out_north')
EnginXFocus(InputWorkspace=ws, DetectorIndices='0-1199', OutputWorkspace='out_indices1')
EnginXFocus(InputWorkspace=ws, DetectorIndices='0-100, 101-500, 400-1199', OutputWorkspace='out_indices3')

# you can compare all these like:
CheckWorkspacesMatch('out_1', 'out_north')
CheckWorkspacesMatch('out_1', 'out_indices1')
# etc.

# note the too big index, should produce an error message:
EnginXFocus(InputWorkspace=ws, DetectorIndices='50000-100000', OutputWorkspace='out_nope')

```

Small note: the approach taken here to add the bank aliases (North, South) is specific to ENGIN-X. We could chat one day about having aliases for banks more in general in Mantid. Not sure if there's general demand for that. One practical problem for this is that it seems that the widespread practice is to put bank names in comments.
